client/packages/inventory/src/Stocktake/DetailView/Footer/StatusChangeButton.tsx

```diff
@@ -72,16 +80,61 @@ const useStatusChangeButton = () => {
 
   const onConfirmStatusChange = async () => {
     if (!selectedOption) return null;
-        const result = await mutateAsync({ id, status: selectedOption.value });
+
+    errors.unsetAll();
+    let result;
+    try {
+      result = await mutateAsync({ id, status: selectedOption.value });
+    } catch (e) {
+      return errorNotification(errorMessage(e))();
+    }
+
     if (result.__typename === 'StocktakeNode') {
       return success(t('messages.saved'))();
     }
 
-    switch (result.error.__typename) {
+    const { error } = result;
+
+    // General errors
+    if (error.__typename === 'CannotEditStocktake') {
+      return errorNotification(t('error.not-editable'))();
+    }
+
+    if (error.__typename === 'StocktakeIsLocked') {
+      return errorNotification(t('error.is-locked'))();
+    }
+
+    let stocktakeLineIdsWithErrors: {
+      [id: string]: StocktakeLineError;
+    } = {};
+    // By line errors
+    switch (error.__typename) {
       case 'StockLinesReducedBelowZero':
-        return error(t('error.stocktake-has-stock-reduced-below-zero'))();
+        // StockLinesReducedBelowZero.errors contains an array of StockLineReducedBelowZero which have StockLines
+        // we want to match StocktakeLine ids for those errors
+        stocktakeLineIdsWithErrors = error.errors.reduce((acc, innerError) => {
```

Comment:
this might be easier if `StockLinesReducedBelowZero` had stocktake lines, but it wouldn't be consistent with `StockLineReducedBelowZero` and we wouldn't have current stockline info when navigating to batch edit (I think we want to show more info about this error eventually and would need latest stockline info for that)

Reply:
I was playing around with existing common utility methods (lodash), and it's very nice. It would mean we would need to be familiar with generic utility functionality vs repo specific utility functions. Slightly more code at invocation end when using lodash but less code in general (since helper method would need to deal with generic types).
9e79f407c2f86cf022f4aa56209f5c07a697b670

And the diff between consuming lodash and generics method would be:
```diff
-        // ids = { stockLineId: stocktakeLineId }
-        const ids = mapValues(
-          mapKeys(lines.nodes, line => line.stockLine?.id),
-          'id'
-        );
-        // mappedErrors = { stockLineId: StockLineReducedBelowZero }
-        const mappedErrors = mapKeys(
+        const mappedErrors = mapErrorsForObject(
+          lines.nodes,
           error.errors,
-          line => ids[line.stockLine.id]
+          (error, stocktakeLine) =>
+            error.stockLine.id === stocktakeLine.stockLine.id
         );
 
         errors.setErrors(mappedErrors);
```